README.md
# smartcab_planner
This is my capstone project for Udacity's Machine Learning Engineer Nanodegree, in which I use reinforcement learning (Q-learning) to come up with a planner that learns which actions to perform in order to reach a point in a grid-like world.

This project was heavily inspired by the nanodegree's Project 4, Teach a Smartcab How to Drive. In fact, this project can be seen as using Q-learning to come up with the planner that is provided in that project.


changelog.md
- First version 
  - learning planner designed as two functions, `learning_planner(state, qvals, initial_val)` and `basic_qval_update(qvals, qval_infos, t)`
  - functions used within the `simulate` function
  - `qval_infos` = experience tuple
  - Hardcoded parameters:
    - learning rate = 1.0 / (1 + t)
    - discount rate = 0.5
    - initial values = 0
- Second version
  - World defined in `PlannerWorld` class
  - Planner defined in `Planner` class and random, perfect, and learner subclasses
  - Experience tuple named `experience`
- Third version
  - Optimism in the face of uncertainty: reward initialized as 1 (but goes to 0 when updating value)
  - Softer learning rate decay: 1.0 / log(1 + t)
  - AMAZING RESULTS!


play.py
from get_next_point import get_next_point
from get_reward import get_reward

import pandas as pd
import random
#import time

GRID = (8, 6)
POSSIBLE_ACTIONS = (None, 'forward', 'left', 'right')
N_TRIALS = 100
DEADLINE = 100


def get_random_position(grid_size):
    return tuple([random.choice(range(number)) + 1
                  for number in grid_size])




#print location, destination, heading

trial_results = []

for i in range(N_TRIALS):
    location = get_random_position(GRID)
    destination = get_random_position(GRID)
    while destination == location:
        destination = get_random_position(GRID)        
    axis = random.choice(range(2))
    heading = tuple([0 if i != axis else random.choice([-1, 1])
                    for i in range(2)])
    result = [destination, location, heading]
    reward_sum = 0

    for i in range(DEADLINE):
        action = random.choice(POSSIBLE_ACTIONS)
        location, heading = get_next_point(location, heading, GRID, action)
        reward = get_reward(location, destination, GRID, 100-i)
        reward_sum += reward
        if location == destination:
            break
    result.extend([i+1, reward_sum])
    trial_results.append(result)

df = pd.DataFrame(trial_results)
df.columns = ['destination', 'initial_location', 'initial_heading',
              'n_steps', 'reward_sum']
print df.describe()

get_delta.py
def get_delta(location, destination, grid_size):
    """
    get the horizontal and vertical distance
    between location and destination.
    """
    delta = [0, 0]
    
    for i in range(2):
        # 1st option: destination to the east/south of location
        if destination[i] > location[i]:
            # two possible distances, going east/south or
            # going west/north
            possible_delta = [destination[i] - location[i], 
                              location[i] + grid_size[i] - destination[i]]
            # if the first distance is the smallest, pick it
            if min(possible_delta) == possible_delta[0]:
                delta[i] = possible_delta[0]
            # if seconde distance is the smallest, pick minus it
            # (the agent will have to go west/north to get to a point
            #  to the east/south)
            else:
                delta[i] = -possible_delta[1]
        # 2nd option: destination to the west/north of location
        else:
            # two possible distances, going west/north or
            # going east/south
            possible_delta = [location[i] - destination[i],
                              destination[i] + grid_size[i] - location[i]]
            # if the first distance is the smallest, pick minus it
            # (the agent will have to go west/north)
            if min(possible_delta) == possible_delta[0]:
                delta[i] = -possible_delta[0]
            # if second distance is smallest, pick it
            else:
                delta[i] = possible_delta[1]
    # return a tuple
    return tuple(delta)

planner_world.py
# -*- coding: utf-8 -*-
"""
"""

import pandas as pd
import random

class PlannerWorld():
    
    def __init__(self, grid_size=(8, 6), trials=1000, deadline=100):
        self.grid_size = grid_size
        self.trials = trials
        self.deadline = deadline
    
    def get_delta(self, location, destination):
        """
        get the horizontal and vertical distance
        between location and destination.
        """
        delta = [0, 0]
        
        for i in range(2):
            # 1st option: destination to the east/south of location
            if destination[i] > location[i]:
                # two possible distances, going east/south or
                # going west/north
                possible_delta = [destination[i] - location[i], 
                                  location[i] + self.grid_size[i] - \
                                  destination[i]]
                # if the first distance is the smallest, pick it
                if min(possible_delta) == possible_delta[0]:
                    delta[i] = possible_delta[0]
                # if seconde distance is the smallest, pick minus it
                # (the agent will have to go west/north to get to a point
                #  to the east/south)
                else:
                    delta[i] = -possible_delta[1]
            # 2nd option: destination to the west/north of location
            else:
                # two possible distances, going west/north or
                # going east/south
                possible_delta = [location[i] - destination[i],
                                  destination[i] + self.grid_size[i] - \
                                  location[i]]
                # if the first distance is the smallest, pick minus it
                # (the agent will have to go west/north)
                if min(possible_delta) == possible_delta[0]:
                    delta[i] = -possible_delta[0]
                # if second distance is smallest, pick it
                else:
                    delta[i] = possible_delta[1]
        # return a tuple
        return tuple(delta)
        
    def get_distance(self, delta):
        """
        Returns the overall distance.
        """
        return sum([abs(i) for i in delta])

    def get_reward(self, steps_left):
        """
        Returns a negative reward proportional to distance if
        the agent has not reached its destination, and a negative
        reward proportional to the time left if the agent has
        reached its destination.
        """
        delta = self.get_delta(self.planner.location, 
                               self.planner.destination)
        distance = self.get_distance(delta)
        reward = -distance if distance else steps_left
        return reward
    
    def move_agent(self, action):
        """
        Modifies the position and heading of the agent.
        """
        location = self.planner.location
        heading = self.planner.heading
        next_point = list(location) # to be modified below
        next_heading = list(heading) # to be modified below
        if action == 'forward': # just move according to the heading
            next_point = [(next_point[i] + next_heading[i]) % \
                            self.grid_size[i]
                          for i in range(2)]
        elif action == 'right':
            if heading[0]: #turned to EW axis
                next_point[1] = (next_point[1] + heading[0]) % \
                        self.grid_size[1]
                next_heading = [heading[1], heading[0]]
            else: # turned to NS axis
                next_point[0] = (next_point[0] - heading[1]) % \
                        self.grid_size[0]
                next_heading = [-heading[1], heading[0]]
        elif action == 'left':
            if heading[0]: # turned to EW axis
                next_point[1] = (next_point[1] - heading[0]) % \
                        self.grid_size[1]
                next_heading = [heading[1], -heading[0]]
            else: # turned to NS axis
                next_point[0] = (next_point[0] + heading[1]) % \
                        self.grid_size[0]
                next_heading = [heading[1], heading[0]]
        else:
            raise ValueError('Invalid action.')
        
        # check if agent has gone to the other side of the world
        next_point = [self.grid_size[i] if next_point[i] == 0 \
                        else next_point[i]
                      for i in range(2)]
                          
        # change agent's position and heading
        self.planner.location = tuple(next_point)
        self.planner.heading = tuple(next_heading)
    
    def get_random_position(self):
        """
        Returns a random position in a grid-like world.
        """
        return tuple([random.choice(range(number)) + 1
                      for number in self.grid_size])

    def simulate(self, planner):
        self.planner = planner
        trial_results = [] # to be populated
        t = 0 # to keep track of the number of actions taken by the agent
    
        for i in range(self.trials):
            # define the agent's location, destination and heading
            self.planner.location = self.get_random_position()
            self.planner.destination = self.get_random_position()
            while self.planner.destination == self.planner.location:
                self.planner.destination = self.get_random_position()        
            axis = random.choice(range(2))
            self.planner.heading = tuple(0 if i != axis else \
                                             random.choice([-1, 1])
                                         for i in range(2))

            # initialize the sum of rewards
            reward_sum = 0

            # for each step in the trial
            for i in range(self.deadline):
                t += 1 # count the time
                # get the delta between location and destination
                delta = self.get_delta(self.planner.location, 
                                       self.planner.destination)
                # set up the agent's state
                state = (delta, self.planner.heading)
                # use the planner to come up with an action
                action = self.planner.policy(state)
                # update the agent's location and heading
                self.move_agent(action)
                # get the next state and the reward associated to it
                next_delta = self.get_delta(self.planner.location,
                                            self.planner.destination)                
                next_state = (next_delta, self.planner.heading)
                reward = self.get_reward(self.deadline - i)
                # use the qval function to update the state's Q-value
                experience = (state, action, reward, next_state)
                self.planner.update_qval(experience, t)
                # update the sum of rewards
                reward_sum += reward
                # check if the agent has reached its destination
                if self.planner.location == self.planner.destination:
                    break
            # store the results of the trial
            trial_results.append((i+1, reward_sum))

        # return a dataframe with the trial results and
        # the dict with Q-values
        df = pd.DataFrame(trial_results)
        df.columns = ['n_steps', 'reward_sum']
        return df, self.planner.qvals

get_next_point.py
def get_next_point(position, heading, grid_size, action):
    next_point = list(position)
    next_heading = list(heading)
    if action == 'forward':
        next_point = [(next_point[i] + heading[i]) % grid_size[i]
                      for i in range(2)]
    elif action == 'right':
        if heading[0]: #turned to EW axis
            next_point[1] = (next_point[1] + heading[0]) % grid_size[1]
            next_heading = [heading[1], heading[0]]
        else: # turned to NS axis
            next_point[0] = (next_point[0] - heading[1]) % grid_size[0]
            next_heading = [-heading[1], heading[0]]
    elif action == 'left':
        if heading[0]:
            next_point[1] = (next_point[1] - heading[0]) % grid_size[1]
            next_heading = [heading[1], -heading[0]]
        else:
            next_point[0] = (next_point[0] + heading[1]) % grid_size[0]
            next_heading = [heading[1], heading[0]]
    elif action != None:
        raise ValueError('Invalid action.')
    
    next_point = [grid_size[i] if next_point[i] == 0 else next_point[i]
                  for i in range(2)]

    return tuple(next_point), tuple(next_heading)

planner_agents.py
# -*- coding: utf-8 -*-
"""
Created on Fri Jul 29 13:28:45 2016

@author: zsj7
"""

import random

POSSIBLE_ACTIONS = ('forward', 'left', 'right')

class Planner():
    def __init__(self, initial_val=0, disc_rate=0.5, 
                 learn_rate_func= lambda x: 1.0 / (x + 1)):
        self.initial_val = initial_val
        self.disc_rate = disc_rate
        self.learn_rate_func = learn_rate_func
        self.qvals = {}
        self.possible_actions = POSSIBLE_ACTIONS

    def update_qval(self, experience, t):
        """
        Returns an updated version of the dict
        with Q-values for (state, action) pairs.
        """
        # learning rate decreases over time
        # according to a predefined function
        learn_rate = self.learn_rate_func(t)

        state, action, reward, next_state = experience
        # get the current Q-value for the (state, action) pair    
        current_qval = self.qvals.get((state, action), 0)
        # get the best action for the next state
        # given the information in the qvals dict
        next_action = self.policy(next_state)
        # update the Q-value for the (state, action) pair    
        update_qval = (1-learn_rate) * current_qval + \
        learn_rate * (reward + self.disc_rate * \
            self.qvals.get((next_state, next_action), 
                           self.initial_val))
        self.qvals[(state, action)] = update_qval

class RandomPlanner(Planner):
    
    def policy(self, state):
        return random.choice(self.possible_actions)
 
class PerfectPlanner(Planner):

    def policy(self, state):
        """
        Returns the best action given the agent's
        position relative to its destination as well as
        its heading.
        """
        delta, heading = state
        # if agent is turned to the east/west axis
        if heading[0]:
            # if it needs to go forward, do it
            if delta[0] * heading[0] > 0:
                return 'forward'
            # else check if it needs to go right or left
            elif delta[1] * heading[0] > 0:
                return 'right'
            else:
                return 'left'
        # if agent is turned to the north/south axis
        else:
            # if it needs to go forward, do it
            if delta[1] * heading[1] > 0:
                return 'forward'
            # else check if it needs to go righ or left
            elif delta[0] * heading[1] < 0:
                return 'right'
            else:
                return 'left'

class LearningPlanner(Planner):
    
    def policy(self, state):
        # get the value of each action given the
        # current state
        possible_vals = {action: self.qvals.get((state, action), 
                                                self.initial_val)
                         for action in self.possible_actions}
    
        # find the actions that yield the largest value
        best_actions = [action for action in possible_vals.keys()
                        if possible_vals[action] == \
                            max(possible_vals.values())]
    
        # randomly pick one of the best actions
        return random.choice(best_actions)

run_sim.py
# -*- coding: utf-8 -*-
"""
Created on Fri Jul 29 13:40:50 2016

@author: zsj7
"""

import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np

plt.rcParams['figure.figsize'] = [15, 10]
plt.rcParams['axes.titlesize'] = 18
plt.rcParams['axes.labelsize'] = 16
plt.rcParams['xtick.labelsize'] = 14
plt.rcParams['ytick.labelsize'] = 14
plt.rcParams['legend.fontsize'] = 12

from planner_world import PlannerWorld
from planner_agents import RandomPlanner, PerfectPlanner, LearningPlanner

GRID = (8, 6)
N_TRIALS = 2000
DEADLINE = 100

def learn_rate_func(t):
    return 1.0 / np.log(t + 1)

random_planner = RandomPlanner()
perfect_planner = PerfectPlanner()
final_learner = LearningPlanner(learn_rate_func=learn_rate_func)
original_learner = LearningPlanner()

world = PlannerWorld(GRID, N_TRIALS, DEADLINE)

df_random, qvals_random = world.simulate(random_planner)
df_perfect, qvals_perfect = world.simulate(perfect_planner)
df_final, qvals_final = world.simulate(final_learner)
df_original, qvals_original = world.simulate(original_learner)

results = {'perfect': df_perfect, 'random': df_random, 
           'original_learner': df_original, 'final_learner': df_final}

panel = pd.Panel(results)

for column in df_perfect.columns:
    sns.boxplot(data=panel.minor_xs(column))
    plt.title(column)
    plt.savefig("{}.png".format(column))
    plt.close()

df_learners = pd.concat([df_final['reward_sum'],
                         df_original['reward_sum']],
                        axis=1)
                        
df_learners.columns = ['final_learner', 'original_learner']

panel.minor_xs('reward_sum').rolling(window=100, center=False).mean().plot()
plt.title("sum of rewards per trial")
plt.xlabel('trial number')
plt.ylabel('sum of rewards')
plt.savefig('planners_progress.png')
plt.close()